controllers/nonadmin_controller.go

```diff
@@ -0,0 +1,182 @@
+		func() error {
+			// Setting Deployment selector if a new object is created, as it is immutable
+			if nonAdminDeployment.ObjectMeta.CreationTimestamp.IsZero() {
+				nonAdminDeployment.Spec.Selector = &metav1.LabelSelector{
+					MatchLabels: nonAdminControllerDeploymentLabel,
+				}
+			}
+
+			nonAdminImage := r.getNonAdminImage(&dpa)
+			err := r.buildNonAdminDeployment(nonAdminDeployment, nonAdminImage)
+			if err != nil {
+				return err
+			}
+
+			// Setting controller owner reference on the non admin controller deployment
+			return controllerutil.SetControllerReference(&dpa, nonAdminDeployment, r.Scheme)
+		},
+	)
+
+	if err != nil {
+		// TODO needed?
+		if k8serror.IsInvalid(err) {
+			cause, isStatusCause := k8serror.StatusCause(err, metav1.CauseTypeFieldValueInvalid)
+			if isStatusCause && cause.Field == "spec.selector" {
+				log.Info("Found immutable selector from previous deployment, recreating non admin controller deployment")
+				err := r.Delete(r.Context, nonAdminDeployment)
+				if err != nil {
+					return false, err
+				}
+				return r.ReconcileNonAdminController(log)
+			}
+		}
+
+		return false, err
+	}
+
+	if operation != controllerutil.OperationResultNone {
+		r.EventRecorder.Event(
+			nonAdminDeployment,
+			corev1.EventTypeNormal,
+			"NonAdminDeploymentReconciled",
+			fmt.Sprintf("Non admin controller deployment %s/%s was %s", nonAdminDeployment.Namespace, nonAdminDeployment.Name, operation),
+		)
+	}
+	return true, nil
+}
+
+func (r *DPAReconciler) buildNonAdminDeployment(deploymentObject *appsv1.Deployment, image string) error {
+	nonAdminContainer := corev1.Container{
+		Image:           image,
+		Name:            nonAdminControllerContainer,
+		ImagePullPolicy: corev1.PullAlways,
+	}
+
+	deploymentObject.Spec = appsv1.DeploymentSpec{
+		Selector: deploymentObject.Spec.Selector,
+		Replicas: pointer.Int32(1),
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: nonAdminControllerDeploymentLabel,
+			},
+			Spec: corev1.PodSpec{
+				RestartPolicy:      corev1.RestartPolicyAlways,
+				Containers:         []corev1.Container{nonAdminContainer},
+				ServiceAccountName: common.OADPOperatorServiceAccount,
+			},
+		},
+	}
+
+	return nil
+}
+
+func (r *DPAReconciler) checkNonAdminEnabled(dpa *oadpv1alpha1.DataProtectionApplication) bool {
+	if dpa.Spec.Features != nil && dpa.Spec.Features.EnableNonAdminMode != nil {
+		return *dpa.Spec.Features.EnableNonAdminMode
+	}
+
+	return false
+}
+
+func (r *DPAReconciler) getNonAdminImage(dpa *oadpv1alpha1.DataProtectionApplication) string {
+	unsupportedOverride := dpa.Spec.UnsupportedOverrides[oadpv1alpha1.NonAdminControllerImageKey]
+	if unsupportedOverride != "" {
+		return unsupportedOverride
+	}
+
+	environmentVariable := os.Getenv("RELATED_IMAGE_NON_ADMIN_CONTROLLER")
+	if environmentVariable != "" {
+		return environmentVariable
+	}
+
+	// TODO change
+	return "quay.io/msouzaol/non-admin-controller:latest"
+}
```

Comment:
other things I saw the other controller do that I did not add
- create configMap
- add some labels
- add some env vars to container

do we need any of this now?

Reply:
All the above things were done to transfer some data from user/oadp operator to vsm controller. For nac we might need some things but nothing as of now.